### PR TITLE
Custom abi decoder

### PR DESCRIPTION
**Pull request by LogvinovLeon**
This PR:
* Removed abi-decoder as a dependency
* Reimplments a subset of it in TS
* Handles the case, when the log can't be decoded

### CODE REVIEW

File: src/utils/abi_decoder.ts [hunk 1]
```diff
@@ -0,0 +1,66 @@
+import * as Web3 from 'web3';
+import * as _ from 'lodash';
+import {AbiType, DecodedLogArgs, DecodedArgs} from '../types';
+import * as SolidityCoder from 'web3/lib/solidity/coder.js';
```
Comment (fabioberger) [inline src/utils/abi_decoder.ts]: do you need the `.js` here?

File: src/utils/abi_decoder.ts [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+export class AbiDecoder {
+    private savedABIs: Web3.AbiDefinition[] = [];
+    private methodIDs: {[signatureHash: string]: Web3.EventAbi} = {};
+    constructor(abiArrays: Web3.AbiDefinition[][]) {
+        _.map(abiArrays, this.addABI.bind(this));
+    }
+    public decodeLog(logItem: Web3.LogEntry): DecodedArgs|undefined {
+        const methodID = logItem.topics[0];
```
Comment (fabioberger) [inline src/utils/abi_decoder.ts]: rename this to `methodID`. Our convention is camelCase except for acronyms.

File: src/utils/abi_decoder.ts [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+            _.map(event.inputs, (param: Web3.EventParameter) => {
+                let value;
+                if (param.indexed) {
+                    value = logItem.topics[topicsIndex];
+                    topicsIndex++;
+                } else {
+                    value = decodedData[dataIndex];
+                    dataIndex++;
+                }
+                if (param.type === 'address') {
```
Comment (fabioberger) [inline src/utils/abi_decoder.ts]: Is this an `ABIType`? If so, let's use that type rather then string literal. Same goes for the subsequent ones.

File: src/utils/abi_decoder.ts [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+                const signature = `${abi.name}(${_.map(abi.inputs, input => input.type).join(',')})`;
+                const signatureHash = new Web3().sha3(signature);
+                this.methodIDs[signatureHash] = abi;
+            }
+        });
+        this.savedABIs = this.savedABIs.concat(abiArray);
+    }
+    private padZeros(address: string) {
+      let formatted = address;
+      if (formatted.indexOf('0x') !== -1) {
```
Comment (fabioberger) [inline src/utils/abi_decoder.ts]: Let's use `!_.includes(formatted, '0x')` instead of `indexOf`

File: src/utils/abi_decoder.ts [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+        formatted = formatted.slice(2);
+      }
+
+      formatted = _.padStart(formatted, 40, '0');
+      return '0x' + formatted;
```
Comment (fabioberger) [inline src/utils/abi_decoder.ts]: template string instead of concatenation.

File: src/utils/abi_decoder.ts (hunk 3 above)
Comment (LogvinovLeon) [reply src/utils/abi_decoder.ts]: It's not an `ABIType`.
```
export enum AbiType {
    Function = 'function',
    Constructor = 'constructor',
    Event = 'event',
    Fallback = 'fallback',
}
```
it's rather `SolidityType`, which we don't have defined yet and I'm rather lazy to define all the `ufixed128x19`s.

**Pull request merged**